### PR TITLE
MDEV-34836: TOI on parent table must BF abort SR in progress on a child

### PR DESCRIPTION
**Pull request by denis-protivensky**
Applied SR transaction on the child table was not BF aborted by TOI running on the parent table for several reasons:
Although SR correctly collected FK-referenced keys to parent, TOI in Galera disregards common certification index and simply sets itself to depend on the latest certified write set seqno.
Since this write set was the fragment of SR transaction, TOI was allowed to run in parallel with SR presuming it would BF abort the latter.

At the same time, DML transactions in the server don't grab MDL locks on FK-referenced tables, thus parent table wasn't protected by an MDL lock from SR and it couldn't provoke MDL lock conflict for TOI to BF abort SR transaction. In InnoDB, DDL transactions grab shared MDL locks on child tables, which is not enough to trigger MDL conflict in Galera. InnoDB-level Wsrep patch didn't contain correct conflict resolution logic due to the fact that it was believed MDL locking should always produce conflicts correctly.

The fix brings conflict resolution rules similar to MDL-level checks to InnoDB, thus accounting for the problematic case.

Apart from that, `wsrep_thd_is_SR()` is patched to return `true` only for executing SR transactions. It should be safe as any other SR state is either the same as for any single write set (thus making the two logically equivalent), or it reflects an SR transaction as being aborting or prepared, which is handled separately in BF-aborting logic, and for regular execution path it should not matter at all.

### CODE REVIEW

**Review by sciascid — CHANGES_REQUESTED** [review]

File: mysql-test/suite/galera/t/MDEV-34836.test [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+INSERT INTO child (parent_id) VALUES (1),(2),(3);
+
+--connection node_1
+# Sync wait for SR transaction to replicate and apply fragments, thus
+# locking parent table as well.
+SET SESSION wsrep_sync_wait = 15;
+SELECT COUNT(*) FROM child;
+# Now run TOI on the parent, which BF-aborts the SR-transaction in progress.
+ALTER TABLE parent AUTO_INCREMENT = 100;
+
```
[the remark this answers is not in the document]
Comment (sciascid) [reply mysql-test/suite/galera/t/MDEV-34836.test:48]: Also, better it would be good to move this test to galera_sr suite.

File: storage/innobase/lock/lock0lock.cc [hunk 2]
```diff
@@ -1052,21 +1076,17 @@ static void lock_wait_wsrep(trx_t *trx)
...
+        // There is no need to kill BF victim with compatible lock.
+        if (wsrep_thd_is_BF(lock->trx->mysql_thd, false) &&
+            !lock_has_to_wait(trx->lock.wait_lock, lock))
         {
-          // There is no need to kill victim with compatible lock
-          if (!lock_has_to_wait(trx->lock.wait_lock, lock))
-            continue;
-
-#ifdef UNIV_DEBUG
-          wsrep_report_error(lock, trx);
```
[the remark this answers is not in the document]
Comment (denis-protivensky) [reply storage/innobase/lock/lock0lock.cc:1068]: Well, if we bring semantics of wsrep transactions in InnoDB, there's no way to simply report BF-BF error for every such a case, because now it may be BF waiting on SR to complete or in prepared state, or BF waiter to be running SR by itself, thus error output produced by wsrep_report_error() becomes false-positive.

Although the current fix is also incorrect, I will need to think more of it.
The problem right now is how to merge wsrep transaction checks with lock_has_to_wait() checks, and also report proper BF-BF error cases.

File: sql/service_wsrep.cc [hunk 3]
```diff
@@ -183,7 +183,21 @@ extern "C" my_bool wsrep_thd_is_BF(const THD *thd, my_bool sync)
 
 extern "C" my_bool wsrep_thd_is_SR(const THD *thd)
 {
-  return thd && thd->wsrep_cs().transaction().is_streaming();
+  return thd && thd->wsrep_cs().transaction().is_streaming() &&
+    thd->wsrep_cs().transaction().state() == wsrep::transaction::s_executing;
+}
+
+extern "C" my_bool wsrep_thd_is_XA_prepared(const MYSQL_THD thd)
```
Comment (sciascid) [inline sql/service_wsrep.cc:190]: In my opinion, this part should be dropped. If you want to fix it for the XA case you need to grab a copy of mariadb enterprise, so that you can test it and fix it there.

File: storage/innobase/lock/lock0lock.cc [hunk 4]
```diff
@@ -495,16 +495,63 @@ void lock_sys_t::close()
...
+
+  /* Applying SR cannot BF abort other high priority (BF). */
+  if (wsrep_thd_is_SR(request_thd))
+      return 1;
+
+  /* Requester is truly BF and granted is applying SR in progress. */
+  if (wsrep_thd_is_SR(granted_thd))
+      return 0;
+
+  /* BF-BF wait. */
```
Comment (sciascid) [inline storage/innobase/lock/lock0lock.cc:542]: Maybe you could move the error reporting here, to avoid -1 / 0 / 1 return codes?

File: sql/service_wsrep.cc [hunk 5]
```diff
@@ -183,7 +183,21 @@ extern "C" my_bool wsrep_thd_is_BF(const THD *thd, my_bool sync)
 
 extern "C" my_bool wsrep_thd_is_SR(const THD *thd)
 {
-  return thd && thd->wsrep_cs().transaction().is_streaming();
+  return thd && thd->wsrep_cs().transaction().is_streaming() &&
+    thd->wsrep_cs().transaction().state() == wsrep::transaction::s_executing;
```
Comment (sciascid) [inline sql/service_wsrep.cc:187]: Can you explain why this change was needed?

File: storage/innobase/lock/lock0lock.cc [hunk 6]
```diff
@@ -1090,16 +1137,22 @@ static void lock_wait_wsrep(trx_t *trx)
...
         {
-          if (wsrep_thd_is_BF(lock->trx->mysql_thd, false))
+          const int has_to_wait= wsrep_BF_has_to_wait(lock, trx);
+          if (!has_to_wait)
           {
-            // There is no need to kill victim with compatible lock
-            if (!lock_has_to_wait(trx->lock.wait_lock, lock))
+            /* There is no need to kill BF victim with compatible lock. */
+            if (wsrep_thd_is_BF(lock->trx->mysql_thd, false) &&
+                !lock_has_to_wait(trx->lock.wait_lock, lock))
```
Comment (sciascid) [inline storage/innobase/lock/lock0lock.cc:1147]: If victim is BF, wsrep_BF_has_to_wait() may be called twice for the same victim. 
The first time, it will called few lines above. It may be called a second time when diving into lock_has_to_wait().
This might get quickly get confusing, if we move some error reporting there, as suggested in another comment.
I wonder if it would be better to change check first if the lock belongs to BF thread? Something along these lines:
```
if (wsrep_thd_is_BF(lock->trx->mysql_thd) {
  if (!lock_has_to_wait()) continue;
} else {
  if (!wsrep_BF_has_to_wait(lock, trx)) continue;
}
```

Comment (denis-protivensky) [reply storage/innobase/lock/lock0lock.cc:1147]: My intent was to preserve the original code as much as possible. But if you think we can change it, then I'll give my vision of it.

First of all, `lock_wait_wsrep()` that does all BF handling has a misleading name. This function is a direct analog to `wsrep_handle_mdl_conflict()`, because its sole purpose is to BF abort conflicting lock holders. So It should be renamed to `wsrep_handle_lock_conflict()`. Then, what's its logic? To detect a conflict between wsrep transactions and then to check if this conflict is real by caling `lock_has_to_wait()`. So why don't we just call `lock_has_to_wait()` in the first place and decide BF actions based on the returned result? Considering that `lock_has_to_wait()` already calls `wsrep_BF_has_to_wait()`, I think it should do all the heavy lifting, including BF-BF checks and aborts, so that the code in `lock_wsrep_wait()` becomes like:

```
if (lock_has_to_wait(..) {
 // BF-abort
}
```
And that's it.

File: storage/innobase/lock/lock0lock.cc (hunk 4 above)
Comment (denis-protivensky) [reply storage/innobase/lock/lock0lock.cc:542]: Yes, it's possible if we move all the BF detection logic to `lock_has_to_wait()`, specifically, to `lock_rec_has_to_wait_wsrep()`. See my comment below.

File: sql/service_wsrep.cc (hunk 5 above)
Comment (denis-protivensky) [reply sql/service_wsrep.cc:187]: BF detection logic is only interested in executing SR transactions, as any other state (maybe except prepared) is a transition state, a "technical" state, and in those states SR transaction is no different from any other one-shot transaction (single write set). We don't want to BF abort SR transactions in any other state than executing. The same is true for other parts where SR transactions are checked: during transaction execution, we're only interested if it's a running SR transaction, not in aborting or in prepared or in committing or in any other state.

From this standpoint, I would argue that SR transaction is either executing or otherwise it's semantically equivalent to a single write set transaction. That's why it makes sense to call something an SR transaction only when it's in the executing state.

Comment (sciascid) [reply sql/service_wsrep.cc:187]: wsrep_thd_is_SR() is also used to decide whether to call wsrep_handle_SR_rollback() during rollback processing. In that context, a transaction is streaming if it has replicated one or more fragments. If that's the case, we want to send a rollback fragment before releasing innodb locks.
Since this is in rollback path (any kind of rollback be it voluntary, bf abort or native deadlock), I'm afraid that the transaction could be in s_must_abort state for example. Failing to send rollback fragment before releasing locks has caused challenging bugs in the past.

Comment (denis-protivensky) [reply sql/service_wsrep.cc:187]: I checked this code path before, and from what I gather `wsrep_handle_SR_rollback()` solely performs BF abort on an SR transaction, which in turn calls `streaming_rollback()`, but only for SR transaction in executing state.
I also checked more wsrep-lib code to ensure that `streaming_rollback()` is called from only three places:
- `certify_fragment()` during an error path
- `before_rollback()`, which doesn't relate to `wsrep_handle_SR_rollback()` code
- `bf_abort()`, which is mentioned earlier

While I agree that such changes need more judging, I think "catch-all" style coding when it's hard to reason about particular states, and which produces false-positives like erroneous debug messages or attempts to perform actions when they should not be performed, is even worse. I do think that trying to BF abort SR transaction in a wrong state issuing wrong debug messages is bad, especially that it erodes understanding what should be called when and under which invariant.

If this specific case is buggy and produced problems in the past, I would expect it to be covered with MTR tests, but I don't see any failures at the moment.

Comment (sciascid) [reply sql/service_wsrep.cc:187]: > [...] which in turn calls streaming_rollback(), but only for SR transaction in executing state.
thanks for clarifying this, I missed it when checking the code path

File: storage/innobase/lock/lock0lock.cc [hunk 7]
```diff
@@ -1051,24 +1099,9 @@ static void lock_wait_wsrep(trx_t *trx)
          (2) If is in XA PREPARE state, it would eventually be rolled
          back and the lock conflict would be resolved when an XA COMMIT
          or XA ROLLBACK statement is executed in some other connection.
-
-         If victim has also BF status, but has earlier seqno, we have to wait.
       */
-      if (lock->trx != trx && lock->trx->mysql_thd &&
```
Comment (sciascid) [inline storage/innobase/lock/lock0lock.cc:1057]: The `lock->trx != trx` clause has disappeared. Was it intentional?

File: storage/innobase/lock/lock0lock.cc [hunk 8]
```diff
@@ -1090,21 +1123,8 @@ static void lock_wait_wsrep(trx_t *trx)
            record-locks instead of table locks. See details
            from comment above.
         */
-        if (lock->trx != trx && lock->trx->mysql_thd &&
-            !(wsrep_thd_is_BF(lock->trx->mysql_thd, false) &&
-              wsrep_thd_order_before(lock->trx->mysql_thd, trx->mysql_thd)))
+        if (lock->trx->mysql_thd && wsrep_will_BF_abort(lock, trx))
```
Comment (sciascid) [inline storage/innobase/lock/lock0lock.cc:1126]: Same as above, the `lock->trx != trx` clause has disappeared.

File: storage/innobase/lock/lock0lock.cc (hunk 7 above)
Comment (denis-protivensky) [reply storage/innobase/lock/lock0lock.cc:1057]: Yes, the check became excessive. If the victim is not BF, then it's a different trx. Otherwise the check for trx equal to itself is in lock_has_to_wait().

**Review by temeo — CHANGES_REQUESTED** [review]

File: storage/innobase/lock/lock0lock.cc [hunk 9]
```diff
@@ -573,8 +557,83 @@ static void wsrep_assert_no_bf_bf_wait(const lock_t *lock, const trx_t *trx,
...
+@param lock  other waiting lock
+@param trx   BF transaction requesting conflicting lock
+*/
+static bool wsrep_will_BF_abort(const lock_t *lock, const trx_t *trx)
+{
+  /* Assume requester is BF. */
+  ut_ad(wsrep_thd_is_BF(trx->mysql_thd, false));
+
+  /* Lock holder is not BF - can BF abort it. */
+  if (!wsrep_thd_is_BF(lock->trx->mysql_thd, false))
```
[the remark this answers is not in the document]
Comment (denis-protivensky) [reply storage/innobase/lock/lock0lock.cc:628]: Yes, it is. Frankly, I put it here because I don't know if it's possible to see lock->trx->is_wsrep() ==  false in this code, so I put an assert in wsrep_BF_has_to_wait() instead and filter non-wsrep trxs here as part of this check, just because almost the same check existed before. Do you know if non-wsrep transactions are possible and if we should BF abort them?

[the remark this answers is not in the document]
Comment (temeo) [reply storage/innobase/lock/lock0lock.cc:628]: I don't expect to see non-wsrep transactions here (maybe a debug assertion would be in place at the beginning of the function). However, wsrep transactions which are *not BF* will always return here, without a lock wait check.

The same check seems to exist in `wsrep_BF_has_to_wait()`, maybe it would be enough?

[the remark this answers is not in the document]
Comment (denis-protivensky) [reply storage/innobase/lock/lock0lock.cc:628]: Yes, I will try an assert instead.

**Review by temeo — APPROVED** [review]
Looks good.
